File: 237G_Computational/Projects/Project1/ComputationalFinance_Project1/Project1/Question4.cpp
#include "Question1.h"
#include "Question4.h"
#include <math.h>
#include <algorithm>
using namespace std;

void Question4(int N, double lambda, double seed)
{
	vector<double> expNums = BuildExponDist(N, lambda, seed);

	//Calculate Probability
	int validCount1 = count_if(expNums.begin(), expNums.end(),
		[&](int num) { return num >= 1; });
	double prob1 = (double)validCount1 / (double)expNums.size();

	int validCount2 = count_if(expNums.begin(), expNums.end(),
		[&](int num) { return num >= 4; });
	double prob2 = (double)validCount2 / (double)expNums.size();

	cout << endl << "QUESTION 4" << endl;
	cout << "Probability that exponential distribution is greater than 1 " << prob1 << endl;
	cout << "Probability that exponential distribution is greater than 4 " << prob2 << endl;

	//Plot Histogram
	PlotHistogram(expNums,false);

	//Mean,Variance
	double mean_exp = mean(expNums);
	double sd_exp = stddev(expNums);
	cout << "Mean of exponential distribution is " << mean_exp << endl;
	cout << "Std Deviation of exponential distribution is " << sd_exp << endl;
}

vector<double> BuildExponDist(int N, double lamda, double seed)
{
	vector<double> finalVals;

	vector<double> randNums = GenerateUniformDist(N, seed);
	for (int count = 0; count < randNums.size(); count++)
	{
		finalVals.push_back(-1.0 * lamda * log(1 - randNums[count]));
	}

	return finalVals;

}

File: 237G_Computational/Projects/Project1/Python/Question4.py
from Question1 import generate_uniform_dist
import math
import matplotlib.pyplot as plot
import numpy as np


def question4(n, lambd, seed):
    exp_nums = build_exponential_distribution(n, lambd, seed)

    prob_greater_1 = len([num for num in exp_nums if num >= 1]) / len(exp_nums)
    prob_greater_4 = len([num for num in exp_nums if num >= 4]) / len(exp_nums)

    print("\n QUESTION 4")
    print("Probability that distribution is greater than 1 {0:.4f}".format(prob_greater_1))
    print("Probability that distribution is greater than 4 {0:.4f}".format(prob_greater_4))

    plot.figure()
    plot.hist(exp_nums)
    plot.title("Exponential distribution histogram")
    plot.ylabel("Frequency")
    plot.xlabel("Value")

    print("Mean of exponential distribution = {0:.4f}".format(np.mean(exp_nums)))
    print("Std dev of exponential distribution = {0:.4f}".format(np.std(exp_nums)))


def build_exponential_distribution(n, lambd, seed):
    rand_nums = generate_uniform_dist(n, seed)
    exp_dist = [-lambd * math.log(1 - val) for val in rand_nums]
    return exp_dist

